### PR TITLE
planner: propagate NDV of column groups across plan nodes

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

First step to solve https://github.com/pingcap/tidb/issues/17185

Problem Summary:

Join size estimation is far from reality.

### What is changed and how it works?

What's Changed:

- Collect column groups which are interested by plan nodes from top-down.
- Maintain NDV of those column groups from bottom-up.

### Related changes

- Need to cherry-pick to the release branch: let's see its effect first.

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- Propagate NDV of column groups across plan nodes

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -463,13 +543,20 @@ func (ds *DataSource) buildIndexMergeOrPath(partialPaths []*util.AccessPath, cur
 }
 
 // DeriveStats implement LogicalPlan DeriveStats interface.
-func (p *LogicalSelection) DeriveStats(childStats []*property.StatsInfo, selfSchema *expression.Schema, childSchema []*expression.Schema) (*property.StatsInfo, error) {
+func (p *LogicalSelection) DeriveStats(childStats []*property.StatsInfo, selfSchema *expression.Schema, childSchema []*expression.Schema, _ [][]*expression.Column) (*property.StatsInfo, error) {
+	if p.stats != nil {
+		return p.stats, nil
+	}
 	p.stats = childStats[0].Scale(SelectionFactor)
+	p.stats.GroupNDVs = nil
```
Comment (qw4990) [inline planner/core/stats.go:551]: How about using the selectivity to adjust the NDV like what you do in `StatsInfo.Scale(factor)`?

File: planner/core/plan.go [hunk 2]
```diff
@@ -166,12 +166,15 @@ type LogicalPlan interface {
...
+	recursiveDeriveStats(colGroups [][]*expression.Column) (*property.StatsInfo, error)
 
 	// DeriveStats derives statistic info for current plan node given child stats.
 	// We need selfSchema, childSchema here because it makes this method can be used in
 	// cascades planner, where LogicalPlan might not record its children or schema.
-	DeriveStats(childStats []*property.StatsInfo, selfSchema *expression.Schema, childSchema []*expression.Schema) (*property.StatsInfo, error)
+	DeriveStats(childStats []*property.StatsInfo, selfSchema *expression.Schema, childSchema []*expression.Schema, colGroups [][]*expression.Column) (*property.StatsInfo, error)
+
+	// ExtractColGroups extracts column groups whose DNV are required by this operator from child.
+	ExtractColGroups(colGroups [][]*expression.Column) [][]*expression.Column
```
Comment (qw4990) [inline planner/core/plan.go:177]: Could you please add more comments to this method? It's better to add some examples...

File: planner/core/stats.go (hunk 1 above)
Comment (qw4990) [reply planner/core/stats.go:551]: Why set it to Nil here? It seems reasonable to adjust(scale) the NDV by `SelectionFactor`.

Comment (eurekaka) [reply planner/core/stats.go:551]: Unlike the DataSource case, the magic number `0.8` here utilizes no information from the filter at all, scaling the GroupNDV with this SelectionFactor may incur estimation errors IMHO, so I reset the GroupNDVs to empty for selection operator.

**Pull request merged**